File: 백준/Gold/1339. 단어 수학/단어 수학.py
import sys 
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
n = int(input())
#가장 자릿수가 큰 숫자가 : 가장 큰 수를 가져감
#자릿수가 동일하다면 : 많이나오는게 가져감
strings_list = ['' for _ in range(n)]
for i in range(n):
    strings_list[i] = input().strip()
dict = {}
for word in strings_list:
    cnt = len(word)-1
    for w in word:
        if w not in dict:
            dict[w] = pow(10, cnt)
        else:
            dict[w] += pow(10, cnt)
        cnt -= 1
dict = sorted(dict.values(), reverse=True)
result = 0
num = 9
for v in dict:
    result += v*num
    num -= 1
print(result)

File: 백준/Gold/1339. 단어 수학/단어 수학.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strings"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	var n int
	var res, num float64
	fmt.Fscanf(reader, "%d\n", &n)
	var numbers = make([]string, n)
	var numMap = make(map[string]float64)
	for i := 0; i < n; i++ {
		fmt.Fscanf(reader, "%s\n", &numbers[i])
	}
	for _, str := range numbers {
		strArr := strings.Split(str, "")
		cnt := float64(len(strArr)) - 1
		for _, s := range strArr {
			if _, exists := numMap[s]; !exists {
				numMap[s] = math.Pow(10, cnt)
			} else {
				numMap[s] += math.Pow(10, cnt)
			}
			cnt -= 1
		}
	}
	var values = []float64{}
	for _, v := range numMap {
		values = append(values, v)
	}
	sort.Sort(sort.Reverse(sort.Float64Slice(values)))
	num = 9
	for _, v := range values {
		res += v * num
		num--
	}
	fmt.Printf("%0.f\n", res)
}